{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "source": [
    "# pycl 1.0.dev1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___\n",
    "**pycl is a package written in python3 containing a collection of generic functions and classes for file parsing, manipulation...**\n",
    "___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pycl contains many functions organized in several categories:\n",
    "\n",
    "* jupyter notebook specific tools\n",
    "* file predicates\n",
    "* path manipulation\n",
    "* string formatting\n",
    "* file manipulation\n",
    "* file information/parsing\n",
    "* directory manipulation\n",
    "* shell manipulation\n",
    "* dictionnary formatting\n",
    "* table formatting\n",
    "* web tools\n",
    "* functions tools\n",
    "* ssh tools\n",
    "\n",
    "Many of the function replicate bash commands in pure python.\n",
    "\n",
    "Please be aware that pycl is an experimental package that is still under development. It was tested under Linux Ubuntu 16.04 and in an HPC environment running under Red Hat Enterprise 7.1. You are welcome to raise issues, contribute to the development and submit patches or updates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ideally, before installation, create a clean python3 virtual environment to deploy the package, using virtualenvwrapper for example (see http://www.simononsoftware.com/virtualenv-tutorial-part-2/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 1: Direct installation with pip from github"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Install the package with pip."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T14:27:42.893738",
     "start_time": "2017-06-21T14:27:42.880250Z"
    },
    "collapsed": false,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "pip3 install git+https://github.com/a-slide/pycl.git"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T10:17:38.416959",
     "start_time": "2017-06-21T10:17:38.404258Z"
    }
   },
   "source": [
    "To update the package:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "pip3 install git+https://github.com/a-slide/pycl.git --upgrade"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 2: Clone the repository and install locally in develop mode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this option, the package will be locally installed in “editable” or “develop” mode. This allows the package to be both installed and editable in project form. This is the recommended option if you wish to participate to the development of the package. As for the option before, the required dependencies will be automatically installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T10:14:57.092212",
     "start_time": "2017-06-21T10:14:57.087096Z"
    },
    "collapsed": false,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "git clone https://github.com/a-slide/pycl.git\n",
    "cd pycl\n",
    "chmod u+x setup.py\n",
    "pip3 install -e ./"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With this option you can also run the testing notebook located in the source directory *pycl/test_pycl.ipynb*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Option 3: Local installation without pip "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This option is also suitable if you are interested in further developing the package, but requires a little bit more hands-on."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Clone the repository locally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "run_control": {
     "frozen": false,
     "read_only": false
    }
   },
   "outputs": [],
   "source": [
    "git clone https://github.com/a-slide/pycl.git"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Add the package directory (./pycl/pycl) to you python3 PATH (depending on you OS and whether you want it to be permanent ot not)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Installing dependencies (optional)\n",
    "\n",
    "By default pycl will not request any third party dependencies during installation, instead dependencies are called on a function basis when needed. If a dependency is missing, the function will display an error message indicating the missing dependency.\n",
    "If one plan to use all the function in pycl they can install the following dependencies:\n",
    "* pandas 0.20.0+\n",
    "* httplib2 0.9.0+\n",
    "* paramiko 2.0.0\n",
    "* notebook 4.0.0+"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using pycl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import pycl main class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T14:50:47.857790",
     "start_time": "2017-06-21T14:50:47.854539Z"
    },
    "collapsed": false,
    "run_control": {
     "frozen": false,
     "read_only": false
    },
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from pycl import pycl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sample data files are provided with the package for testing purpose. Depending on your installation choice, they could be found either directly in the *pycl/data* directory or using *pkg_resources* (see https://setuptools.readthedocs.io/en/latest/pkg_resources.html). Here is the list of the test files:\n",
    "\n",
    "* Small_m5C_Squires_hg38.bed\n",
    "* gencode_sample.gff3\n",
    "* RADAR_clean.txt\n",
    "* RADAR_Secondary.txt.gz\n",
    "* Small_m5C_Squires_hg38_reformat.bed\n",
    "* stdout.txt\n",
    "* Small_editing_Peng_hg38_reformat.bed\n",
    "* RADAR_Secondary.txt\n",
    "* RADAR_Main.txt\n",
    "* Small_editing_Peng_hg38.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T14:47:54.171361",
     "start_time": "2017-06-21T14:47:54.160809Z"
    },
    "collapsed": false,
    "run_control": {
     "frozen": false,
     "read_only": false
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Small_m5C_Squires_hg38.bed\n",
      "gencode_sample.gff3\n",
      "RADAR_clean.txt\n",
      "RADAR_Secondary.txt.gz\n",
      "Small_m5C_Squires_hg38_reformat.bed\n",
      "stdout.txt\n",
      "Small_editing_Peng_hg38_reformat.bed\n",
      "RADAR_Secondary.txt\n",
      "RADAR_Main.txt\n",
      "Small_editing_Peng_hg38.bed\n"
     ]
    }
   ],
   "source": [
    "# Cell to generate the list of files. To be removed in the README.md \n",
    "from pkg_resources import Requirement, resource_filename\n",
    "from os import listdir\n",
    "for file in listdir(resource_filename(Requirement.parse(\"pycl\"),'pycl/data/')):\n",
    "    print (file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A list of the available functions is provided below. Functions are comprehensively detailed in the testing notebook provided with the package or in html version on nbviewer: [link to test_notebook](https://nbviewer.jupyter.org/github/a-slide/pycl/blob/master/pycl/test_pycl.ipynb?flush_cache=true)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** bash **\n",
    "\n",
    "bash(cmd, live='stdout', print_stdout=True, ret_stdout=False, log_stdout=None, print_stderr=True, ret_stderr=False, log_stderr=None)\n",
    "\n",
    "More advanced version of bash calling with live printing of the standard output and possibilities to log the redirect\n",
    "the output and error as a string return or directly in files. If ret_stderr and ret_stdout are True a tuple will be returned and if\n",
    "both are False None will be returned\n",
    "* cmd A command line string formatted as a string\n",
    "* print_stdout\n",
    "If True the standard output will be LIVE printed through the system standard output stream\n",
    "* ret_stdout\n",
    "If True the standard output will be returned as a string\n",
    "* log_stdout\n",
    "If a filename is given, the standard output will logged in this file\n",
    "* print_stderr\n",
    "If True the standard error will be printed through the system standard error stream\n",
    "* ret_stderr\n",
    "If True the standard error will be returned as a string\n",
    "* log_stderr\n",
    "If a filename is given, the standard error will logged in this file\n",
    "\n",
    "________________________________________________________________________________\n",
    "** bash_basic **\n",
    "\n",
    "bash_basic(cmd)\n",
    "\n",
    "Sent basic bash command\n",
    "\n",
    "________________________________________________________________________________\n",
    "** bash_update **\n",
    "\n",
    "bash_update(cmd, update_freq=1)\n",
    "FOR JUPYTER NOTEBOOK\n",
    "Run a bash command and print the output in the cell. The output is updated each time until the output is None.\n",
    "This is suitable for monitoring tasks that log events until there is nothing else to print such as bjobs or bpeeks.\n",
    "* cmd\n",
    "    A command line string formatted as a string\n",
    "* update_freq\n",
    "    The frequency of output updating in seconds [DEFAULT: 1]\n",
    "\n",
    "________________________________________________________________________________\n",
    "** cat **\n",
    "\n",
    "cat(fp, max_lines=100, line_numbering=False)\n",
    "\n",
    "Emulate linux cat cmd but with line cap protection. Handle gziped files\n",
    "\n",
    "________________________________________________________________________________\n",
    "** colsum **\n",
    "\n",
    "colsum(fp, colrange=None, separator='', header=False, ignore_hashtag_line=False, max_items=10, ret_type='md')\n",
    "\n",
    "Create a summary of selected columns of a file\n",
    "* fp\n",
    "    Path to the file to be parsed\n",
    "* colrange\n",
    "    A list of column index to parse\n",
    "* separator\n",
    "    A character or a list of characters to split the lines\n",
    "* ignore_hashtag_line\n",
    "    skip line starting with a # symbol\n",
    "* max_items\n",
    "    maximum item per line\n",
    "* ret_type\n",
    "    Possible return types:\n",
    "    md = markdown formatted table,\n",
    "    dict = raw parsing dict,\n",
    "    report = Indented_text_report\n",
    "\n",
    "________________________________________________________________________________\n",
    "** copyFile **\n",
    "\n",
    "copyFile(src, dest)\n",
    "\n",
    "Copy a single file to a destination file or folder (with error handling/reporting)\n",
    "* src\n",
    "    Source file path\n",
    "* dest\n",
    "    Path of the folder where to copy the source file\n",
    "\n",
    "________________________________________________________________________________\n",
    "** count_uniq **\n",
    "\n",
    "count_uniq(fp, colnum, select_values=None, drop_values=None, skip_comment='#', sep='\\t')\n",
    "\n",
    "Count unique occurences in a specific column of a tabulated file\n",
    "* fp\n",
    "    Path to the file to be parsed (gzipped or not)\n",
    "* colnum\n",
    "    Index number of the column to summarize\n",
    "* select_values\n",
    "    Select specific lines in the file based on a dictionary containing column index(es) and valu(es) or list\n",
    "    of values to select. Exemple {2:[\"exon\", \"transcript\"], 4:\"lincRNA\"}. DEFAULT=None\n",
    "* drop_values\n",
    "    Same think that select_value but will drop the lines instead. DEFAULT=None\n",
    "* skip_comment\n",
    "    Drop any comment lines starting with this character. DEFAULT=\"#\"\n",
    "* sep\n",
    "    Character or list of characters to use in order to split the lines. Exemple [\"  \",\";\"]. DEFAULT=\"       \"\n",
    "\n",
    "________________________________________________________________________________\n",
    "** dict_to_md **\n",
    "\n",
    "dict_to_md(d, key_label='', value_label='', transpose=False, sort_by_key=False, sort_by_val=True, max_items=None)\n",
    "\n",
    "Def to transform a dict into a markdown formated table\n",
    "\n",
    "________________________________________________________________________________\n",
    "** dict_to_report **\n",
    "\n",
    "dict_to_report(d, tab='\\t', ntab=0, sep=':', sort_dict=True, max_items=None)\n",
    "\n",
    "Recursive function to return a text report from nested dict or OrderedDict objects\n",
    "\n",
    "________________________________________________________________________________\n",
    "** dir_name **\n",
    "\n",
    "dir_name(path)\n",
    "\n",
    "Return the complete path where is located the file without the file name\n",
    "\n",
    "________________________________________________________________________________\n",
    "** fastcount **\n",
    "\n",
    "fastcount(fp)\n",
    "\n",
    "Efficient way to count the number of lines in a file. Handle gziped files\n",
    "\n",
    "________________________________________________________________________________\n",
    "** file_basename **\n",
    "\n",
    "file_basename(path)\n",
    "\n",
    "Return the basename of a file without folder location and extension\n",
    "\n",
    "________________________________________________________________________________\n",
    "** file_extension **\n",
    "\n",
    "file_extension(path)\n",
    "\n",
    "Return The extension of a file in lower-case. If archived file (\"gz\", \"zip\", \"xz\", \"bz2\")\n",
    "the method will output the base extension + the archive extension\n",
    "\n",
    "________________________________________________________________________________\n",
    "** file_name **\n",
    "\n",
    "file_name(path)\n",
    "\n",
    "Return The complete name of a file with the extension but without folder location\n",
    "\n",
    "________________________________________________________________________________\n",
    "** gunzip_file **\n",
    "\n",
    "gunzip_file(in_path, out_path=None)\n",
    "\n",
    "ungzip a file\n",
    "* in_path\n",
    "    Path of the input compressed file\n",
    "* out_path\n",
    "    Path of the output uncompressed file (facultative)\n",
    "\n",
    "________________________________________________________________________________\n",
    "** gzip_file **\n",
    "\n",
    "gzip_file(in_path, out_path=None)\n",
    "\n",
    "gzip a file\n",
    "* in_path\n",
    "    Path of the input uncompressed file\n",
    "* out_path\n",
    "    Path of the output compressed file (facultative)\n",
    "\n",
    "________________________________________________________________________________\n",
    "** head **\n",
    "\n",
    "head(fp, n=10, line_numbering=False, ignore_hashtag_line=False)\n",
    "\n",
    "Emulate linux head cmd. Handle gziped files\n",
    "\n",
    "________________________________________________________________________________\n",
    "** is_gziped **\n",
    "\n",
    "is_gziped(fp)\n",
    "\n",
    "Return True if the file is Gziped else False\n",
    "\n",
    "________________________________________________________________________________\n",
    "** is_readable_file **\n",
    "\n",
    "is_readable_file(fp)\n",
    "\n",
    "Verify the readability of a file or list of file\n",
    "\n",
    "________________________________________________________________________________\n",
    "** jprint **\n",
    "\n",
    "jprint(*args, **kwargs)\n",
    "\n",
    "FOR JUPYTER NOTEBOOK ONLY\n",
    "Format a string in HTML and print the output. Equivalent of print, but highly customizable\n",
    "Many options can be passed to the function.\n",
    "* args\n",
    "    One or several objects that can be cast in str\n",
    "* kwargs\n",
    "    Formatting options to tweak the html rendering\n",
    "    Boolean options : bold, italic, highlight, underlined, striked, subscripted, superscripted\n",
    "    String oprions: font, color, size, align, background_color, line_height\n",
    "\n",
    "________________________________________________________________________________\n",
    "** larger_display **\n",
    "\n",
    "larger_display(percent=100)\n",
    "\n",
    "FOR JUPYTER NOTEBOOK ONLY\n",
    "Resize the area of the screen containing the notebook according to a given percentage of the available width\n",
    "*  percent percentage of the width of the screen to use [DEFAULT:100]\n",
    "\n",
    "________________________________________________________________________________\n",
    "** linerange **\n",
    "\n",
    "linerange(fp, range_list=[], line_numbering=True)\n",
    "\n",
    "Print a range of lines in a file according to a list of start end lists. Handle gziped files\n",
    "* fp\n",
    "    Path to the file to be parsed\n",
    "* range_list\n",
    "    list of start, end coordinates lists or tuples\n",
    "* line_numbering\n",
    "    If True the number of the line will be indicated in front of the line\n",
    "\n",
    "________________________________________________________________________________\n",
    "** make_cmd_str **\n",
    "\n",
    "make_cmd_str(prog_name, opt_dict={}, opt_list=[])\n",
    "\n",
    "Create a Unix like command line string from the prog name, a dict named arguments and a list of unmammed arguments\n",
    "exemple make_cmd_str(\"bwa\", {\"b\":None, t\":6, \"i\":\"../idx/seq.fa\"}, [\"../read1\", \"../read2\"])\n",
    "* prog_name\n",
    "    Name (if added to the system path) or path of the program\n",
    "* opt_dict\n",
    "    Dictionary of option arguments such as \"-t 5\". The option flag have to be the key (without \"-\") and the the option value in the\n",
    "    dictionary value. If no value is requested after the option flag \"None\" had to be assigned to the value field.\n",
    "* opt_list\n",
    "    List of simple command line arguments\n",
    "\n",
    "________________________________________________________________________________\n",
    "** mkdir **\n",
    "\n",
    "mkdir(fp, level=1)\n",
    "\n",
    "Reproduce the ability of UNIX \"mkdir -p\" command\n",
    "(ie if the path already exits no exception will be raised).\n",
    "Can create nested directories by recursivity\n",
    "* fp    path name where the folder should be created\n",
    "*level    level in the path where to start to create the directories. Used by the program for the recursive creation of directories\n",
    "\n",
    "________________________________________________________________________________\n",
    "** print_arg **\n",
    "\n",
    "print_arg()\n",
    "\n",
    "Print calling function named and unnamed arguments\n",
    "\n",
    "________________________________________________________________________________\n",
    "** reformat_table **\n",
    "\n",
    "reformat_table(input_file, output_file='', return_df=False, init_template=[], final_template=[], header='', keep_original_header=True, header_from_final_template=False, replace_internal_space='_', replace_null_val='*', subst_dict={}, filter_dict=[], predicate=None, standard_template=None, verbose=False)\n",
    "\n",
    "Reformat a table given an initial and a final line templates indicated as a list where numbers\n",
    "indicate the data column and strings the formatting characters\n",
    "\n",
    "*  input_file\n",
    "    A file with a structured text formatting (gzipped or not)\n",
    "*  output_file\n",
    "    A file path to output the reformatted table (if empty will not write in a file)\n",
    "*  return_df\n",
    "    If true will return a pandas dataframe containing the reformated table (Third party pandas package required) by default the columns\n",
    "    will be names after the final template [DEFAULT:False]\n",
    "*  init_template\n",
    "    A list of indexes and separators describing the structure of the input file\n",
    "        Example initial line = \"chr1    631539    631540    Squires|id1    0    +\"\n",
    "        Initial template = [0,\"     \",1,\"   \",2,\"   \",3,\"|\",4,\"     \",5,\"   \",6]\n",
    "        Alternatively, instead of the numbers, string indexes can be used, but they need to be enclosed in curly brackets to\n",
    "        differentiate them from the separators. This greatly simplify the writing of the final template.\n",
    "        Example initial line = \"chr1    631539    631540    Squires|id1    0    +\"\n",
    "        Initial template = [\"{chrom}\",\"     \",\"{start}\",\"   \",\"{end}\",\"|\",\"{name}\",\"        \",\"{score}\",\"   \",\"{strand}\"]\n",
    "*  final_template\n",
    "    A list of indexes and separators describing the required structure of the output file. Name indexes need to match indexes of the\n",
    "    init_template and have to follow the same synthax  [DEFAULT:Same that init template]\n",
    "        Example final line = \"chr1    631539    631540    m5C|-|HeLa|22344696    -    -\"\n",
    "        Final template = [0,\"       \",1,\"   \",2,\"   m5C|-|HeLa|22344696     -       \",6]\n",
    "*  header\n",
    "    A string to write as a file header at the beginning of the file\n",
    "*  keep_original_header\n",
    "    If True the original header of the input file will be copied at the beginning of the output file [DEFAULT:True]\n",
    "*  header_from_final_template\n",
    "    Generate a header according to the name or number of the fields given in the final_template [DEFAULT:True]\n",
    "*  replace_internal_space\n",
    "    All internal blank space will be replaced by this character [DEFAULT:\"_\"]\n",
    "*  replace_null_val\n",
    "    Field with no value will be replaced by this character [DEFAULT:\"*\"]\n",
    "*  subst_dict\n",
    "    Nested dictionary of substitution per position to replace specific values by others [DEFAULT:None]\n",
    "        Example: { 0:{\"chr1\":\"1\",\"chr2\":\"2\"}, 3:{\"Squires\":\"5376774764\",\"Li\":\"27664684\"}}\n",
    "*  filter_dict\n",
    "    A dictionary of list per position  to filter out lines  with specific values [DEFAULT:None]\n",
    "        Example: { 0:[\"chr2\", \"chr4\"], 1:[\"46767\", \"87765\"], 5:[\"76559\", \"77543\"]}\n",
    "*  predicate\n",
    "    A lambda predicate function for more advance filtering operations [DEFAULT:None]\n",
    "        Example:  lambda val_dict: abs(int(val_dict[1])-int(val_dict[2])) <= 2000\n",
    "*  standard_template\n",
    "    Existing standard template to parse the file  instead of providing one manually. List of saved templates:\n",
    "    - \"gff3_ens_gene\" = Template for ensembl gff3 fields. Select only the genes lines and decompose to individual elements.\n",
    "    - \"gff3_ens_transcript\" = Template for ensembl gff3 fields. Select only the transcript lines and decompose to individual elements.\n",
    "    - \"gtf_ens_gene\" = Template for ensembl gft fields. Select only the genes lines and decompose to individual elements\n",
    "* verbose\n",
    "    If True will print detailed information [DEFAULT:False]\n",
    "\n",
    "________________________________________________________________________________\n",
    "** rm_blank **\n",
    "\n",
    "rm_blank(name, replace='')\n",
    "\n",
    "Replace blank spaces in a name by a given character (default = remove)\n",
    "Blanks at extremities are always removed and nor replaced\n",
    "\n",
    "________________________________________________________________________________\n",
    "** scp **\n",
    "\n",
    "scp(hostname, local_file, remote_dir, username=None, rsa_private_key=None, ssh_config='~/.ssh/config')\n",
    "\n",
    "Copy a file over ssh in a target remote directory\n",
    "* hostname\n",
    "    Name of the host ssh server\n",
    "* username\n",
    "    name of the user\n",
    "* rsa_private_key\n",
    "    path to the rsa private key\n",
    "* local_file\n",
    "    path to the local file\n",
    "* remote_dir\n",
    "    path to the target directory\n",
    "* ssh_config\n",
    "    use as an alternative method instead of giving the username and rsa_private_key. Will fetch them from the config file directly\n",
    "________________________________________________________________________________\n",
    "** simplecount **\n",
    "\n",
    "simplecount(fp, ignore_hashtag_line=False)\n",
    "\n",
    "Simple way to count the number of lines in a file with more options\n",
    "________________________________________________________________________________\n",
    "** supersplit **\n",
    "\n",
    "supersplit(string, separator='')\n",
    "\n",
    "like split but can take a list of separators instead of a simple separator\n",
    "\n",
    "________________________________________________________________________________\n",
    "** tail **\n",
    "\n",
    "tail(fp, n=10, line_numbering=False)\n",
    "\n",
    "Emulate linux tail cmd. Handle gziped files\n",
    "\n",
    "________________________________________________________________________________\n",
    "** toogle_code **\n",
    "\n",
    "toogle_code()\n",
    "\n",
    "FOR JUPYTER NOTEBOOK ONLY\n",
    "Hide code with a clickable link in a jupyter notebook\n",
    "\n",
    "________________________________________________________________________________\n",
    "** url_exist **\n",
    "\n",
    "url_exist(url)\n",
    "Predicate verifying if an url exist without downloading all the link\n",
    "\n",
    "________________________________________________________________________________\n",
    "** wget **\n",
    "\n",
    "wget(url, out_name='', progress_block=100000000)\n",
    "\n",
    "Download a file from an URL to a local storage.\n",
    "*  url\n",
    "    A internet URL pointing to the file to download\n",
    "*  outname\n",
    "    Name of the outfile where (facultative)\n",
    "*  progress_block\n",
    "    size of the byte block for the progression of the download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T15:56:56.647017",
     "start_time": "2017-06-21T15:56:56.620998Z"
    },
    "collapsed": false,
    "run_control": {
     "frozen": false,
     "read_only": false
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**bash**\n",
      "Help on function bash in module pycl.pycl:\n",
      "\n",
      "bash(cmd, live='stdout', print_stdout=True, ret_stdout=False, log_stdout=None, print_stderr=True, ret_stderr=False, log_stderr=None)\n",
      "    More advanced version of bash calling with live printing of the standard output and possibilities to log the redirect\n",
      "    the output and error as a string return or directly in files. If ret_stderr and ret_stdout are True a tuple will be returned and if\n",
      "    both are False None will be returned\n",
      "    * cmd A command line string formatted as a string\n",
      "    * print_stdout\n",
      "        If True the standard output will be LIVE printed through the system standard output stream\n",
      "    * ret_stdout\n",
      "        If True the standard output will be returned as a string\n",
      "    * log_stdout\n",
      "        If a filename is given, the standard output will logged in this file\n",
      "    * print_stderr\n",
      "        If True the standard error will be printed through the system standard error stream\n",
      "    * ret_stderr\n",
      "        If True the standard error will be returned as a string\n",
      "    * log_stderr\n",
      "        If a filename is given, the standard error will logged in this file\n",
      "\n",
      "________________________________________________________________________________\n",
      "**bash_basic**\n",
      "Help on function bash_basic in module pycl.pycl:\n",
      "\n",
      "bash_basic(cmd)\n",
      "    Sent basic bash command\n",
      "\n",
      "________________________________________________________________________________\n",
      "**bash_update**\n",
      "Help on function bash_update in module pycl.pycl:\n",
      "\n",
      "bash_update(cmd, update_freq=1)\n",
      "    FOR JUPYTER NOTEBOOK\n",
      "    Run a bash command and print the output in the cell. The output is updated each time until the output is None.\n",
      "    This is suitable for monitoring tasks that log events until there is nothing else to print such as bjobs or bpeeks.\n",
      "    * cmd\n",
      "        A command line string formatted as a string\n",
      "    * update_freq\n",
      "        The frequency of output updating in seconds [DEFAULT: 1]\n",
      "\n",
      "________________________________________________________________________________\n",
      "**cat**\n",
      "Help on function cat in module pycl.pycl:\n",
      "\n",
      "cat(fp, max_lines=100, line_numbering=False)\n",
      "    Emulate linux cat cmd but with line cap protection. Handle gziped files\n",
      "\n",
      "________________________________________________________________________________\n",
      "**colsum**\n",
      "Help on function colsum in module pycl.pycl:\n",
      "\n",
      "colsum(fp, colrange=None, separator='', header=False, ignore_hashtag_line=False, max_items=10, ret_type='md')\n",
      "    Create a summary of selected columns of a file\n",
      "    * fp\n",
      "        Path to the file to be parsed\n",
      "    * colrange\n",
      "        A list of column index to parse\n",
      "    * separator\n",
      "        A character or a list of characters to split the lines\n",
      "    * ignore_hashtag_line\n",
      "        skip line starting with a # symbol\n",
      "    * max_items\n",
      "        maximum item per line\n",
      "    * ret_type\n",
      "        Possible return types:\n",
      "        md = markdown formatted table,\n",
      "        dict = raw parsing dict,\n",
      "        report = Indented_text_report\n",
      "\n",
      "________________________________________________________________________________\n",
      "**copyFile**\n",
      "Help on function copyFile in module pycl.pycl:\n",
      "\n",
      "copyFile(src, dest)\n",
      "    Copy a single file to a destination file or folder (with error handling/reporting)\n",
      "    * src\n",
      "        Source file path\n",
      "    * dest\n",
      "        Path of the folder where to copy the source file\n",
      "\n",
      "________________________________________________________________________________\n",
      "**count_uniq**\n",
      "Help on function count_uniq in module pycl.pycl:\n",
      "\n",
      "count_uniq(fp, colnum, select_values=None, drop_values=None, skip_comment='#', sep='\\t')\n",
      "    Count unique occurences in a specific column of a tabulated file\n",
      "    * fp\n",
      "        Path to the file to be parsed (gzipped or not)\n",
      "    * colnum\n",
      "        Index number of the column to summarize\n",
      "    * select_values\n",
      "        Select specific lines in the file based on a dictionary containing column index(es) and valu(es) or list\n",
      "        of values to select. Exemple {2:[\"exon\", \"transcript\"], 4:\"lincRNA\"}. DEFAULT=None\n",
      "    * drop_values\n",
      "        Same think that select_value but will drop the lines instead. DEFAULT=None\n",
      "    * skip_comment\n",
      "        Drop any comment lines starting with this character. DEFAULT=\"#\"\n",
      "    * sep\n",
      "        Character or list of characters to use in order to split the lines. Exemple [\"  \",\";\"]. DEFAULT=\"       \"\n",
      "\n",
      "________________________________________________________________________________\n",
      "**dict_to_md**\n",
      "Help on function dict_to_md in module pycl.pycl:\n",
      "\n",
      "dict_to_md(d, key_label='', value_label='', transpose=False, sort_by_key=False, sort_by_val=True, max_items=None)\n",
      "    Def to transform a dict into a markdown formated table\n",
      "\n",
      "________________________________________________________________________________\n",
      "**dict_to_report**\n",
      "Help on function dict_to_report in module pycl.pycl:\n",
      "\n",
      "dict_to_report(d, tab='\\t', ntab=0, sep=':', sort_dict=True, max_items=None)\n",
      "    Recursive function to return a text report from nested dict or OrderedDict objects\n",
      "\n",
      "________________________________________________________________________________\n",
      "**dir_name**\n",
      "Help on function dir_name in module pycl.pycl:\n",
      "\n",
      "dir_name(path)\n",
      "    Return the complete path where is located the file without the file name\n",
      "\n",
      "________________________________________________________________________________\n",
      "**fastcount**\n",
      "Help on function fastcount in module pycl.pycl:\n",
      "\n",
      "fastcount(fp)\n",
      "    Efficient way to count the number of lines in a file. Handle gziped files\n",
      "\n",
      "________________________________________________________________________________\n",
      "**file_basename**\n",
      "Help on function file_basename in module pycl.pycl:\n",
      "\n",
      "file_basename(path)\n",
      "    Return the basename of a file without folder location and extension\n",
      "\n",
      "________________________________________________________________________________\n",
      "**file_extension**\n",
      "Help on function file_extension in module pycl.pycl:\n",
      "\n",
      "file_extension(path)\n",
      "    Return The extension of a file in lower-case. If archived file (\"gz\", \"zip\", \"xz\", \"bz2\")\n",
      "    the method will output the base extension + the archive extension\n",
      "\n",
      "________________________________________________________________________________\n",
      "**file_name**\n",
      "Help on function file_name in module pycl.pycl:\n",
      "\n",
      "file_name(path)\n",
      "    Return The complete name of a file with the extension but without folder location\n",
      "\n",
      "________________________________________________________________________________\n",
      "**gopen**\n",
      "Help on function open in module gzip:\n",
      "\n",
      "open(filename, mode='rb', compresslevel=9, encoding=None, errors=None, newline=None)\n",
      "    Open a gzip-compressed file in binary or text mode.\n",
      "    \n",
      "    The filename argument can be an actual filename (a str or bytes object), or\n",
      "    an existing file object to read from or write to.\n",
      "    \n",
      "    The mode argument can be \"r\", \"rb\", \"w\", \"wb\", \"x\", \"xb\", \"a\" or \"ab\" for\n",
      "    binary mode, or \"rt\", \"wt\", \"xt\" or \"at\" for text mode. The default mode is\n",
      "    \"rb\", and the default compresslevel is 9.\n",
      "    \n",
      "    For binary mode, this function is equivalent to the GzipFile constructor:\n",
      "    GzipFile(filename, mode, compresslevel). In this case, the encoding, errors\n",
      "    and newline arguments must not be provided.\n",
      "    \n",
      "    For text mode, a GzipFile object is created, and wrapped in an\n",
      "    io.TextIOWrapper instance with the specified encoding, error handling\n",
      "    behavior, and line ending(s).\n",
      "\n",
      "________________________________________________________________________________\n",
      "**gunzip_file**\n",
      "Help on function gunzip_file in module pycl.pycl:\n",
      "\n",
      "gunzip_file(in_path, out_path=None)\n",
      "    ungzip a file\n",
      "    * in_path\n",
      "        Path of the input compressed file\n",
      "    * out_path\n",
      "        Path of the output uncompressed file (facultative)\n",
      "\n",
      "________________________________________________________________________________\n",
      "**gzip_file**\n",
      "Help on function gzip_file in module pycl.pycl:\n",
      "\n",
      "gzip_file(in_path, out_path=None)\n",
      "    gzip a file\n",
      "    * in_path\n",
      "        Path of the input uncompressed file\n",
      "    * out_path\n",
      "        Path of the output compressed file (facultative)\n",
      "\n",
      "________________________________________________________________________________\n",
      "**head**\n",
      "Help on function head in module pycl.pycl:\n",
      "\n",
      "head(fp, n=10, line_numbering=False, ignore_hashtag_line=False)\n",
      "    Emulate linux head cmd. Handle gziped files\n",
      "\n",
      "________________________________________________________________________________\n",
      "**is_gziped**\n",
      "Help on function is_gziped in module pycl.pycl:\n",
      "\n",
      "is_gziped(fp)\n",
      "    Return True if the file is Gziped else False\n",
      "\n",
      "________________________________________________________________________________\n",
      "**is_readable_file**\n",
      "Help on function is_readable_file in module pycl.pycl:\n",
      "\n",
      "is_readable_file(fp)\n",
      "    Verify the readability of a file or list of file\n",
      "\n",
      "________________________________________________________________________________\n",
      "**jprint**\n",
      "Help on function jprint in module pycl.pycl:\n",
      "\n",
      "jprint(*args, **kwargs)\n",
      "    FOR JUPYTER NOTEBOOK ONLY\n",
      "    Format a string in HTML and print the output. Equivalent of print, but highly customizable\n",
      "    Many options can be passed to the function.\n",
      "    * args\n",
      "        One or several objects that can be cast in str\n",
      "    ** kwargs\n",
      "        Formatting options to tweak the html rendering\n",
      "        Boolean options : bold, italic, highlight, underlined, striked, subscripted, superscripted\n",
      "        String oprions: font, color, size, align, background_color, line_height\n",
      "\n",
      "________________________________________________________________________________\n",
      "**larger_display**\n",
      "Help on function larger_display in module pycl.pycl:\n",
      "\n",
      "larger_display(percent=100)\n",
      "    FOR JUPYTER NOTEBOOK ONLY\n",
      "    Resize the area of the screen containing the notebook according to a given percentage of the available width\n",
      "    *  percent percentage of the width of the screen to use [DEFAULT:100]\n",
      "\n",
      "________________________________________________________________________________\n",
      "**linerange**\n",
      "Help on function linerange in module pycl.pycl:\n",
      "\n",
      "linerange(fp, range_list=[], line_numbering=True)\n",
      "    Print a range of lines in a file according to a list of start end lists. Handle gziped files\n",
      "    * fp\n",
      "        Path to the file to be parsed\n",
      "    * range_list\n",
      "        list of start, end coordinates lists or tuples\n",
      "    * line_numbering\n",
      "        If True the number of the line will be indicated in front of the line\n",
      "\n",
      "________________________________________________________________________________\n",
      "**make_cmd_str**\n",
      "Help on function make_cmd_str in module pycl.pycl:\n",
      "\n",
      "make_cmd_str(prog_name, opt_dict={}, opt_list=[])\n",
      "    Create a Unix like command line string from the prog name, a dict named arguments and a list of unmammed arguments\n",
      "    exemple make_cmd_str(\"bwa\", {\"b\":None, t\":6, \"i\":\"../idx/seq.fa\"}, [\"../read1\", \"../read2\"])\n",
      "    * prog_name\n",
      "        Name (if added to the system path) or path of the program\n",
      "    * opt_dict\n",
      "        Dictionary of option arguments such as \"-t 5\". The option flag have to be the key (without \"-\") and the the option value in the\n",
      "        dictionary value. If no value is requested after the option flag \"None\" had to be assigned to the value field.\n",
      "    * opt_list\n",
      "        List of simple command line arguments\n",
      "\n",
      "________________________________________________________________________________\n",
      "**mkdir**\n",
      "Help on function mkdir in module pycl.pycl:\n",
      "\n",
      "mkdir(fp, level=1)\n",
      "    Reproduce the ability of UNIX \"mkdir -p\" command\n",
      "    (ie if the path already exits no exception will be raised).\n",
      "    Can create nested directories by recursivity\n",
      "    * fp\n",
      "        path name where the folder should be created\n",
      "    *level\n",
      "        level in the path where to start to create the directories. Used by the program for the recursive creation of directories\n",
      "\n",
      "________________________________________________________________________________\n",
      "**print_arg**\n",
      "Help on function print_arg in module pycl.pycl:\n",
      "\n",
      "print_arg()\n",
      "    Print calling function named and unnamed arguments\n",
      "\n",
      "________________________________________________________________________________\n",
      "**reformat_table**\n",
      "Help on function reformat_table in module pycl.pycl:\n",
      "\n",
      "reformat_table(input_file, output_file='', return_df=False, init_template=[], final_template=[], header='', keep_original_header=True, header_from_final_template=False, replace_internal_space='_', replace_null_val='*', subst_dict={}, filter_dict=[], predicate=None, standard_template=None, verbose=False)\n",
      "    Reformat a table given an initial and a final line templates indicated as a list where numbers\n",
      "    indicate the data column and strings the formatting characters\n",
      "    \n",
      "    *  input_file\n",
      "        A file with a structured text formatting (gzipped or not)\n",
      "    *  output_file\n",
      "        A file path to output the reformatted table (if empty will not write in a file)\n",
      "    *  return_df\n",
      "        If true will return a pandas dataframe containing the reformated table (Third party pandas package required) by default the columns\n",
      "        will be names after the final template [DEFAULT:False]\n",
      "    *  init_template\n",
      "        A list of indexes and separators describing the structure of the input file\n",
      "            Example initial line = \"chr1    631539    631540    Squires|id1    0    +\"\n",
      "            Initial template = [0,\"     \",1,\"   \",2,\"   \",3,\"|\",4,\"     \",5,\"   \",6]\n",
      "            Alternatively, instead of the numbers, string indexes can be used, but they need to be enclosed in curly brackets to\n",
      "            differentiate them from the separators. This greatly simplify the writing of the final template.\n",
      "            Example initial line = \"chr1    631539    631540    Squires|id1    0    +\"\n",
      "            Initial template = [\"{chrom}\",\"     \",\"{start}\",\"   \",\"{end}\",\"|\",\"{name}\",\"        \",\"{score}\",\"   \",\"{strand}\"]\n",
      "    *  final_template\n",
      "        A list of indexes and separators describing the required structure of the output file. Name indexes need to match indexes of the\n",
      "        init_template and have to follow the same synthax  [DEFAULT:Same that init template]\n",
      "            Example final line = \"chr1    631539    631540    m5C|-|HeLa|22344696    -    -\"\n",
      "            Final template = [0,\"       \",1,\"   \",2,\"   m5C|-|HeLa|22344696     -       \",6]\n",
      "    *  header\n",
      "        A string to write as a file header at the beginning of the file\n",
      "    *  keep_original_header\n",
      "        If True the original header of the input file will be copied at the beginning of the output file [DEFAULT:True]\n",
      "    *  header_from_final_template\n",
      "        Generate a header according to the name or number of the fields given in the final_template [DEFAULT:True]\n",
      "    *  replace_internal_space\n",
      "        All internal blank space will be replaced by this character [DEFAULT:\"_\"]\n",
      "    *  replace_null_val\n",
      "        Field with no value will be replaced by this character [DEFAULT:\"*\"]\n",
      "    *  subst_dict\n",
      "        Nested dictionary of substitution per position to replace specific values by others [DEFAULT:None]\n",
      "            Example: { 0:{\"chr1\":\"1\",\"chr2\":\"2\"}, 3:{\"Squires\":\"5376774764\",\"Li\":\"27664684\"}}\n",
      "    *  filter_dict\n",
      "        A dictionary of list per position  to filter out lines  with specific values [DEFAULT:None]\n",
      "            Example: { 0:[\"chr2\", \"chr4\"], 1:[\"46767\", \"87765\"], 5:[\"76559\", \"77543\"]}\n",
      "    *  predicate\n",
      "        A lambda predicate function for more advance filtering operations [DEFAULT:None]\n",
      "            Example:  lambda val_dict: abs(int(val_dict[1])-int(val_dict[2])) <= 2000\n",
      "    *  standard_template\n",
      "        Existing standard template to parse the file  instead of providing one manually. List of saved templates:\n",
      "        - \"gff3_ens_gene\" = Template for ensembl gff3 fields. Select only the genes lines and decompose to individual elements.\n",
      "        - \"gff3_ens_transcript\" = Template for ensembl gff3 fields. Select only the transcript lines and decompose to individual elements.\n",
      "        - \"gtf_ens_gene\" = Template for ensembl gft fields. Select only the genes lines and decompose to individual elements\n",
      "    * verbose\n",
      "        If True will print detailed information [DEFAULT:False]\n",
      "\n",
      "________________________________________________________________________________\n",
      "**rm_blank**\n",
      "Help on function rm_blank in module pycl.pycl:\n",
      "\n",
      "rm_blank(name, replace='')\n",
      "    Replace blank spaces in a name by a given character (default = remove)\n",
      "    Blanks at extremities are always removed and nor replaced\n",
      "\n",
      "________________________________________________________________________________\n",
      "**scp**\n",
      "Help on function scp in module pycl.pycl:\n",
      "\n",
      "scp(hostname, local_file, remote_dir, username=None, rsa_private_key=None, ssh_config='~/.ssh/config')\n",
      "    Copy a file over ssh in a target remote directory\n",
      "    * hostname\n",
      "        Name of the host ssh server\n",
      "    * username\n",
      "        name of the user\n",
      "    * rsa_private_key\n",
      "        path to the rsa private key\n",
      "    * local_file\n",
      "        path to the local file\n",
      "    * remote_dir\n",
      "        path to the target directory\n",
      "    * ssh_config\n",
      "        use as an alternative method instead of giving the username and rsa_private_key. Will fetch them from the config file directly\n",
      "\n",
      "________________________________________________________________________________\n",
      "**shutilCopy**\n",
      "Help on function copy in module shutil:\n",
      "\n",
      "copy(src, dst, *, follow_symlinks=True)\n",
      "    Copy data and mode bits (\"cp src dst\"). Return the file's destination.\n",
      "    \n",
      "    The destination may be a directory.\n",
      "    \n",
      "    If follow_symlinks is false, symlinks won't be followed. This\n",
      "    resembles GNU's \"cp -P src dst\".\n",
      "    \n",
      "    If source and destination are the same file, a SameFileError will be\n",
      "    raised.\n",
      "\n",
      "________________________________________________________________________________\n",
      "**simplecount**\n",
      "Help on function simplecount in module pycl.pycl:\n",
      "\n",
      "simplecount(fp, ignore_hashtag_line=False)\n",
      "    Simple way to count the number of lines in a file with more options\n",
      "\n",
      "________________________________________________________________________________\n",
      "**supersplit**\n",
      "Help on function supersplit in module pycl.pycl:\n",
      "\n",
      "supersplit(string, separator='')\n",
      "    like split but can take a list of separators instead of a simple separator\n",
      "\n",
      "________________________________________________________________________________\n",
      "**tail**\n",
      "Help on function tail in module pycl.pycl:\n",
      "\n",
      "tail(fp, n=10, line_numbering=False)\n",
      "    Emulate linux tail cmd. Handle gziped files\n",
      "\n",
      "________________________________________________________________________________\n",
      "**toogle_code**\n",
      "Help on function toogle_code in module pycl.pycl:\n",
      "\n",
      "toogle_code()\n",
      "    FOR JUPYTER NOTEBOOK ONLY\n",
      "    Hide code with a clickable link in a jupyter notebook\n",
      "\n",
      "________________________________________________________________________________\n",
      "**url_exist**\n",
      "Help on function url_exist in module pycl.pycl:\n",
      "\n",
      "url_exist(url)\n",
      "    Predicate verifying if an url exist without downloading all the link\n",
      "\n",
      "________________________________________________________________________________\n",
      "**wget**\n",
      "Help on function wget in module pycl.pycl:\n",
      "\n",
      "wget(url, out_name='', progress_block=100000000)\n",
      "    Download a file from an URL to a local storage.\n",
      "    *  url\n",
      "        A internet URL pointing to the file to download\n",
      "    *  outname\n",
      "        Name of the outfile where (facultative)\n",
      "    *  progress_block\n",
      "        size of the byte block for the progression of the download\n",
      "\n",
      "________________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "# Cell to generate the list of functions. To be removed in the README.md \n",
    "from inspect import getmembers, isfunction\n",
    "for name, type  in getmembers(pycl):\n",
    "    if isfunction(type) and not name.startswith(\"_\"):\n",
    "        print(\"**{}**\".format(name))\n",
    "        help(type)\n",
    "        print(\"_\"*80)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-06-21T10:53:03.615489",
     "start_time": "2017-06-21T10:53:03.601671Z"
    }
   },
   "source": [
    "## Authors and Contact\n",
    "\n",
    "Adrien Leger - 2017\n",
    "\n",
    "Enright's group, EMBL EBI\n",
    "\n",
    "* <aleg@ebi.ac.uk>\n",
    "* [Github](https://github.com/a-slide)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_setup_py3",
   "language": "python",
   "name": "test_setup_py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
